File: shared/src/main/java/com/runtracker/shared/data/repository/SwimmingRepository.kt
package com.runtracker.shared.data.repository

import com.runtracker.shared.data.db.SwimmingWorkoutDao
import com.runtracker.shared.data.db.SwimmingTrainingPlanDao
import com.runtracker.shared.data.model.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import java.util.*
class SwimmingRepository(
    private val workoutDao: SwimmingWorkoutDao,
    private val planDao: SwimmingTrainingPlanDao
) {
    // Workout operations
    fun getAllWorkouts(): Flow<List<SwimmingWorkout>> = workoutDao.getAllWorkouts()
    
    suspend fun getWorkoutById(id: Long): SwimmingWorkout? = workoutDao.getWorkoutById(id)
    
    fun getWorkoutsInRange(startTime: Long, endTime: Long): Flow<List<SwimmingWorkout>> = 
        workoutDao.getWorkoutsInRange(startTime, endTime)
    
    fun getRecentCompletedWorkouts(limit: Int = 10): Flow<List<SwimmingWorkout>> = 
        workoutDao.getRecentCompletedWorkouts(limit)
    
    suspend fun getActiveWorkout(): SwimmingWorkout? = workoutDao.getActiveWorkout()
    
    suspend fun insertWorkout(workout: SwimmingWorkout): Long = workoutDao.insert(workout)
    
    suspend fun updateWorkout(workout: SwimmingWorkout) = workoutDao.update(workout)
    
    suspend fun deleteWorkout(workout: SwimmingWorkout) = workoutDao.delete(workout)
    
    suspend fun getTotalDistanceSince(startTime: Long): Double = 
        workoutDao.getTotalDistanceSince(startTime) ?: 0.0
    
    suspend fun getWorkoutCountSince(startTime: Long): Int = 
        workoutDao.getWorkoutCountSince(startTime)
    
    // Training plan operations
    fun getAllPlans(): Flow<List<SwimmingTrainingPlan>> = planDao.getAllPlans()
    
    fun getActivePlan(): Flow<SwimmingTrainingPlan?> = planDao.getActivePlan()
    
    suspend fun getPlanById(id: Long): SwimmingTrainingPlan? = planDao.getPlanById(id)
    
    suspend fun insertPlan(plan: SwimmingTrainingPlan): Long = planDao.insert(plan)
    
    suspend fun updatePlan(plan: SwimmingTrainingPlan) = planDao.update(plan)
    
    suspend fun deletePlan(plan: SwimmingTrainingPlan) = planDao.delete(plan)
    
    suspend fun activatePlan(id: Long) {
        planDao.deactivateAllPlans()
        planDao.activatePlan(id)
    }
    
    // Get next upcoming workout from active plan
    suspend fun getNextUpcomingWorkout(): Pair<ScheduledSwimWorkout, Int>? {
        val plan = getActivePlan().first() ?: return null
        
        val calendar = Calendar.getInstance()
        val today = calendar.get(Calendar.DAY_OF_WEEK)
        val startDate = Calendar.getInstance().apply { timeInMillis = plan.startDate }
        val currentWeek = ((calendar.timeInMillis - startDate.timeInMillis) / (7 * 24 * 60 * 60 * 1000)).toInt() + 1
        
        // Find next workout from today onwards
        for (daysAhead in 0..60) {
            val checkDate = Calendar.getInstance().apply { add(Calendar.DAY_OF_YEAR, daysAhead) }
            val checkDayOfWeek = checkDate.get(Calendar.DAY_OF_WEEK)
            val checkWeek = ((checkDate.timeInMillis - startDate.timeInMillis) / (7 * 24 * 60 * 60 * 1000)).toInt() + 1
            
            val workout = plan.weeklySchedule.find { 
                it.weekNumber == checkWeek && 
                it.dayOfWeek == checkDayOfWeek && 
                !it.isCompleted &&
                it.workoutType != SwimWorkoutType.REST_DAY
            }
            
            if (workout != null) {
                return Pair(workout, daysAhead)
            }
        }
        
        return null
    }
    
    // Mark workout as completed
    suspend fun markWorkoutCompleted(planId: Long, workoutId: String, completedWorkoutId: Long) {
        val plan = getPlanById(planId) ?: return
        val updatedSchedule = plan.weeklySchedule.map { workout ->
            if (workout.id == workoutId) {
                workout.copy(isCompleted = true, completedWorkoutId = completedWorkoutId)
            } else {
                workout
            }
        }
        updatePlan(plan.copy(weeklySchedule = updatedSchedule))
    }
    
    // Calculate weekly stats
    suspend fun getWeeklyStats(): SwimmingWeeklyStats {
        val calendar = Calendar.getInstance()
        calendar.set(Calendar.DAY_OF_WEEK, calendar.firstDayOfWeek)
        calendar.set(Calendar.HOUR_OF_DAY, 0)
        calendar.set(Calendar.MINUTE, 0)
        calendar.set(Calendar.SECOND, 0)
        calendar.set(Calendar.MILLISECOND, 0)
        val weekStart = calendar.timeInMillis
        
        val totalDistance = getTotalDistanceSince(weekStart)
        val workoutCount = getWorkoutCountSince(weekStart)
        
        return SwimmingWeeklyStats(
            totalDistanceMeters = totalDistance,
            workoutCount = workoutCount,
            weekStartTimestamp = weekStart
        )
    }
}

data class SwimmingWeeklyStats(
    val totalDistanceMeters: Double,
    val workoutCount: Int,
    val weekStartTimestamp: Long
) {
    val totalDistanceKm: Double get() = totalDistanceMeters / 1000.0
}

File: iosApp/GoSteady/Sources/Repositories/SwimmingRepository.swift
import Foundation
import SwiftData

final class SwimmingRepository {
    private let context: ModelContext

    init(context: ModelContext) {
        self.context = context
    }

    // MARK: - Workouts

    func fetchAllWorkouts() -> [SwimmingWorkout] {
        let descriptor = FetchDescriptor<SwimmingWorkout>(sortBy: [SortDescriptor(\.startTime, order: .reverse)])
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchWorkoutById(_ id: UUID) -> SwimmingWorkout? {
        let descriptor = FetchDescriptor<SwimmingWorkout>(predicate: #Predicate { $0.id == id })
        return try? context.fetch(descriptor).first
    }

    func fetchWorkoutsInRange(start: Date, end: Date) -> [SwimmingWorkout] {
        let descriptor = FetchDescriptor<SwimmingWorkout>(
            predicate: #Predicate { $0.startTime >= start && $0.startTime <= end },
            sortBy: [SortDescriptor(\.startTime, order: .reverse)]
        )
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchRecentCompleted(limit: Int = 10) -> [SwimmingWorkout] {
        var descriptor = FetchDescriptor<SwimmingWorkout>(
            predicate: #Predicate { $0.isCompleted == true },
            sortBy: [SortDescriptor(\.startTime, order: .reverse)]
        )
        descriptor.fetchLimit = limit
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchActiveWorkout() -> SwimmingWorkout? {
        let descriptor = FetchDescriptor<SwimmingWorkout>(predicate: #Predicate { $0.isCompleted == false })
        return try? context.fetch(descriptor).first
    }

    func insert(_ workout: SwimmingWorkout) {
        context.insert(workout)
        try? context.save()
    }

    func update(_ workout: SwimmingWorkout) {
        try? context.save()
    }

    func delete(_ workout: SwimmingWorkout) {
        context.delete(workout)
        try? context.save()
    }

    func totalDistanceSince(_ date: Date) -> Double {
        let workouts = fetchWorkoutsInRange(start: date, end: Date()).filter(\.isCompleted)
        return workouts.reduce(0) { $0 + $1.distanceMeters }
    }

    func workoutCountSince(_ date: Date) -> Int {
        let workouts = fetchWorkoutsInRange(start: date, end: Date()).filter(\.isCompleted)
        return workouts.count
    }

    // MARK: - Training Plans

    func fetchAllPlans() -> [SwimmingTrainingPlan] {
        let descriptor = FetchDescriptor<SwimmingTrainingPlan>(sortBy: [SortDescriptor(\.startDate, order: .reverse)])
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchActivePlan() -> SwimmingTrainingPlan? {
        let descriptor = FetchDescriptor<SwimmingTrainingPlan>(predicate: #Predicate { $0.isActive == true })
        return try? context.fetch(descriptor).first
    }

    func fetchPlanById(_ id: UUID) -> SwimmingTrainingPlan? {
        let descriptor = FetchDescriptor<SwimmingTrainingPlan>(predicate: #Predicate { $0.id == id })
        return try? context.fetch(descriptor).first
    }

    func insertPlan(_ plan: SwimmingTrainingPlan) {
        context.insert(plan)
        try? context.save()
    }

    func updatePlan(_ plan: SwimmingTrainingPlan) {
        try? context.save()
    }

    func deletePlan(_ plan: SwimmingTrainingPlan) {
        context.delete(plan)
        try? context.save()
    }

    func activatePlan(_ plan: SwimmingTrainingPlan) {
        let allPlans = fetchAllPlans().filter(\.isActive)
        for p in allPlans { p.isActive = false }
        plan.isActive = true
        try? context.save()
    }

    func markWorkoutCompleted(planId: UUID, workoutId: String, completedWorkoutId: UUID) {
        guard let plan = fetchPlanById(planId) else { return }
        if let index = plan.weeklySchedule.firstIndex(where: { $0.id == workoutId }) {
            plan.weeklySchedule[index].isCompleted = true
            plan.weeklySchedule[index].completedWorkoutId = completedWorkoutId
            try? context.save()
        }
    }

    // MARK: - Weekly Stats

    func weeklyStats() -> SwimmingWeeklyStats {
        let weekStart = Date().startOfWeek
        let workouts = fetchWorkoutsInRange(start: weekStart, end: Date()).filter(\.isCompleted)
        let totalDistance = workouts.reduce(0.0) { $0 + $1.distanceMeters }
        let count = workouts.count
        let avgPace: Double
        if totalDistance > 0 {
            let totalDuration = workouts.reduce(0.0) { $0 + Double($1.durationMillis) / 1000.0 }
            avgPace = totalDuration / (totalDistance / 100.0)
        } else {
            avgPace = 0
        }
        return SwimmingWeeklyStats(totalDistanceMeters: totalDistance, workoutCount: count, averagePace: avgPace, weekStartDate: weekStart)
    }
}

struct SwimmingWeeklyStats {
    let totalDistanceMeters: Double
    let workoutCount: Int
    let averagePace: Double
    let weekStartDate: Date

    var totalDistanceKm: Double { totalDistanceMeters / 1000.0 }
}
